### PR TITLE
Bouquet of UI enhancements

### PR DESCRIPTION
**Pull request by cristian64**
A number improvements to the UI:

- Add splitter in main window to allow resizing:

- Store window state and window geometry in settings file. State is restored on startup.
- Provide a window title to all dialogs (**Add Watch**, **Edit Watch**, **Memory Viewer**, **Dolphin Memory Engine**, ...).
- Rework **Add Watch** dialog with a proper form layout:

image

- Memory watch list is now saved as part of the settings; list is restored on startup.
- Support for memory addresses that start with the `0x` prefix.
- Rows in the memory watch table are no longer sortable: users could mistakenly reordered their curated list.

Bug fixes:

- Crash involving a double deallocation in `DlgAddWatchEntry`.

### CODE REVIEW

**Comment by cristian64** [comment]
I thought I'd open this PR with a few UI improvements that I wish had been available when I was using this tool.

I'm not expecting much traction, though, so I will abandon the PR after a few days, as I wouldn't like to increase the number of stale PRs.

**Comment by cristian64** [comment]
As promised, no need to contribute to the list of moldering PRs. **Abandoned**.

**Comment by dreamsyntax** [comment]
Reopening as I will be revisiting this / using as a reference.

**Comment by dreamsyntax** [comment]
@cristian64 Would you be willing to rebase this PR? Looked it over, I'm basically fine merging as is once resolved. If you have no interest in this project anymore I can just cherry-pick your commits onto it.

**Comment by cristian64** [comment]
Sounds good to me. Give me a couple of days to rebase.

File: Source/GUI/MainWindow.cpp
Comment (dreamsyntax) [inline Source/GUI/MainWindow.cpp:34]: A few lines above this the firstHookAttempt is already called. Is there a particular reason we should keep this?

**Review by dreamsyntax — CHANGES_REQUESTED** [review]
Sounds good, thank you.

File: Source/GUI/MainWindow.cpp
Comment (cristian64) [reply Source/GUI/MainWindow.cpp:34]: At the moment, `hookIfNotHooked()` schedules _itself_ for another run in 3000 ms, so it needs this first call to start the loop.

This call could be changed to
```cpp
QTimer::singleShot(3000, this, &MainWindow::hookIfNotHooked);
```
, so the loop is started without trying to hook.

I think my preference would be to remove the `firstHookAttempt()` function, and add the `updateMem2Status()` call directly in `hookIfNotHooked()`, and then we can keep this call to `hookIfNotHooked()` here, which then starts the cycle.

**Comment by dreamsyntax** [comment]
If it makes sense to add it here, the ability to disable the watchlist sort would be appreciated. 

Otherwise Ill just open a PR after this is merged.

**Comment by cristian64** [comment]
> If it makes sense to add it here, the ability to disable the watchlist sort would be appreciated. (Clicking the tab column names)
> 
> Otherwise Ill just open a PR after this is merged.

I added a commit with the `setSortingEnabled()` line removed. (I agree it was a bit risky; the user could easily lose the order.)

File: Source/GUI/MainWindow.cpp
Comment (cristian64) [reply Source/GUI/MainWindow.cpp:34]: This has been done.

(I finally remembered why `updateMem2Status()` was not being called in `hookIfNotHooked()`: it's already called in `onHookAttempt()`.)

**Comment by cristian64** [comment]
It should have been rebased now.

Really glad the project has been resuscitated; I still use it almost every day. There are other improvements that I'd like to make, but I guess it can be done in follow-up PRs. Some of these changes are:

- Reduce padding/margins in the UI to fit more useful data. It feels like Qt wastes much space at the moment.
- Remove some of the churn in the GUI, like those big buttons, **Hook** or **Open memory viewer**, that could be moved into the top-bar menu, specially now with the auto-hook mechanism.
- Fix some crashes that occur in certain conditions when Dolphin is closed while DME is hooked.

**Comment by dreamsyntax** [comment]
> It should have been rebased now.
> 
> Really glad the project has been resuscitated; I still use it almost every day. There are other improvements that I'd like to make, but I guess it can be done in follow-up PRs. Some of these changes are:
> 
> - Reduce padding/margins in the UI to fit more useful data. It feels like Qt wastes much space at the moment.
> - Remove some of the churn in the GUI, like those big buttons, **Hook** or **Open memory viewer**, that could be moved into the top-bar menu, specially now with the auto-hook mechanism.
> - Fix some crashes that occur in certain conditions when Dolphin is closed while DME is hooked.

Glad to hear it. I welcome changes, but please hold off Until Qt6 merge (soon). You can base your PR off Qt6 PR if you dont want to wait.

File: Source/GUI/MemWatcher/Dialogs/DlgAddWatchEntry.cpp
Comment (dreamsyntax) [inline Source/GUI/MemWatcher/Dialogs/DlgAddWatchEntry.cpp:44]: I'm not following why this is 10 now instead of 8? Wouldnt the address always be XXXXYYYY ?

File: Source/GUI/MainWindow.cpp
Comment (dreamsyntax) [inline Source/GUI/MainWindow.cpp:174]: Looking at this again, I have some concerns about constantly making a QTimer singleShot. I'll have to profile and test in real world use.

File: Source/GUI/MemWatcher/Dialogs/DlgAddWatchEntry.cpp
Comment (cristian64) [reply Source/GUI/MemWatcher/Dialogs/DlgAddWatchEntry.cpp:44]: It's mentioned in the commit: to allow for the `0x` prefix.

File: Source/GUI/MainWindow.cpp
Comment (dreamsyntax) [reply Source/GUI/MainWindow.cpp:174]: After testing this build in real world use, I absolutely want there to be a setting to **Turn Off** auto hooking.
In my own use case I regularly purposely unhook with multiple DMEs up with different freeze values already pre-loaded.

**Review by dreamsyntax — CHANGES_REQUESTED** [review]

File: Source/GUI/MemWatcher/Dialogs/DlgAddWatchEntry.cpp
Comment (dreamsyntax) [reply Source/GUI/MemWatcher/Dialogs/DlgAddWatchEntry.cpp:44]: Thanks for clarifying.

File: Source/GUI/MainWindow.cpp
Comment (cristian64) [reply Source/GUI/MainWindow.cpp:174]: Performance wise it's not going to be a concern; it wouldn't be different to a `QTimer`. I wouldn't disagree that a `QTimer` may be a more orthodox approach, though.

Comment (cristian64) [reply Source/GUI/MainWindow.cpp:174]: I think it's a sensible request. I was hoping to add options for disabling auto-hooking once the big **Hook** button is moved to a different place, but I can do it now (or take this out of this PR).

Comment (dreamsyntax) [reply Source/GUI/MainWindow.cpp:174]: It depends if you think its easy enough to implement/add.

I would suggest doing something like this (adding to the settings.ini we generate)
https://github.com/aldelaro5/Dolphin-memory-engine/pull/72/files

Comment (cristian64) [reply Source/GUI/MainWindow.cpp:174]: I'll take auto-hooking out, and open a different PR for it.

Comment (cristian64) [reply Source/GUI/MainWindow.cpp]: Commit removed from branch. I will open a different PR shortly (which can then stay on hold until the Qt 6 work is completed).

**Review by dreamsyntax — APPROVED** [review]

**Pull request merged**
